mapadroid/worker/MITMBase.py

```diff
@@ -55,123 +56,83 @@ def __init__(self, args, dev_id, origin, last_known_state, communicator: Abstrac
+                self.logger.debug("Checking GMO location reported by {} at {} against real data location {}",
+                                  self._origin,
+                                  self.current_location,
                                   latest_location)
                 distance_to_data = get_distance_of_two_points_in_meters(float(latest_location.lat),
                                                                         float(latest_location.lng),
                                                                         float(self.current_location.lat),
                                                                         float(self.current_location.lng))
                 max_distance_of_mode = self._mapping_manager.routemanager_get_max_radius(self._routemanager_name)
-                max_distance_for_worker: int = self._applicationArgs.maximum_valid_distance \
-                    if max_distance_of_mode < self._applicationArgs.maximum_valid_distance else max_distance_of_mode
+                max_distance_for_worker = self._applicationArgs.maximum_valid_distance
+                if max_distance_for_worker > max_distance_of_mode > 5:
+                    # some modes may be too strict (e.g. quests with 0.0001m calculations for routes)
+                    # yet, the route may "require" a stricter ruling than max valid distance
+                    max_distance_for_worker = max_distance_of_mode
+
                 self.logger.debug("Distance of worker {} to data location: {}", self._origin, distance_to_data)
-                if max_distance_for_worker and distance_to_data > max_distance_for_worker:
-                    self.logger.debug("Real data too far from worker position, waiting...")
+                if distance_to_data > max_distance_for_worker:
+                    self.logger.debug("Real data too far from worker position, waiting, max distance allowed: {}m",
+                                      max_distance_for_worker)
                     check_data = False
+            elif latest_location is not None and latest_location.lat == latest_location.lng == 1000:
+                self.logger.warning("Data may be valid but does not contain a proper location yet.")
+                check_data = False
+            elif proto_to_wait_for == 106 and latest_location is None:
+                # just wait for the next GMO to get a location...
+                check_data = False
 
             if check_data:
-                data_requested = self._wait_data_worker(
+                type_of_data_returned, data = self._check_for_data_content(
                     latest, proto_to_wait_for, timestamp)
+
             if not self._mapping_manager.routemanager_present(self._routemanager_name) \
                     or self._stop_worker_event.is_set():
                 self.logger.error("killed while sleeping")
                 raise InternalStopWorkerException
+            if type_of_data_returned == LatestReceivedType.UNDEFINED:
```

Comment:
This may be better if its a loop instead of a sleep since we are not guaranteed of a correct result